### PR TITLE
[bugfix] Clean Validation Sanity Checking metrics

### LINKED ISSUE #8162: Metrics logged during sanity checking still appears in training
## 🐛 Bug

Metrics logged during sanity checking still appears in training.

## Reproduction using the BoringModel
The code logs in `validation_step` and `validation_epoch_end` only when `self.trainer.sanity_checking` is `True`, but the following progress bar is outputted. This demonstrates leakage of the logging from the sanity checking to the training.
`Epoch 0: 100%
626/626 [00:01<00:00, 546.35it/s, loss=1.54e-14, v_num=2, val_step_metric_epoch=42.00, val_epoch_end_metric=1337.0`

https://colab.research.google.com/drive/1IQSw9AuoK5OH-NYHrk5rhX2Aa7bylSgg?usp=sharing

### Expected behavior
Metrics logged during sanity checking should be cleared or not logged at all.
Progress bar, logger and callback metrics should be empty when the training loop starts.

### Environment
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.3.7post0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR cleans the metrics generated during sanity checking.

Fixes #8162

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 1]
```diff
@@ -712,3 +712,35 @@ def training_step(self, *args):
...
+        def validation_step(self, batch, batch_idx):
+            output = super().validation_step(batch, batch_idx)
+            if self.trainer.sanity_checking:
+                self.log("val_loss", output["x"], prog_bar=True, logger=True)
+            return output
+
+        def training_step(self, batch, batch_idx):
+            loss = super().training_step(batch, batch_idx)
+            if batch_idx == 0:
+                pass
```
Comment (awaelchli) [inline tests/trainer/logging_/test_train_loop_logging.py:730]:
```suggestion
            if batch_idx == 0:
```
probably better to remove it 😄

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 2]
```diff
@@ -712,3 +712,35 @@ def training_step(self, *args):
     model = TestModel()
     with pytest.raises(MisconfigurationException, match=r'reduce_fx={min,max,mean,sum}\)` are currently supported'):
         trainer.fit(model)
+
+
+def test_sanity_metrics_are_reset_and_dont_appear_in_prog_bar(tmpdir):
```
Comment (awaelchli) [inline tests/trainer/logging_/test_train_loop_logging.py:717]: great! the test fails on master. However, I tried setting `num_sanity_val_steps=0` just to be sure but we get an error:

```
tmpdir = local('/private/var/folders/s2/338zrsts1b17v5cj18xr8_640000gn/T/pytest-of-adrian/pytest-108/test_sanity_metrics_are_reset_0')

    def test_sanity_metrics_are_reset_and_dont_appear_in_prog_bar(tmpdir):
    
        class TestModel(BoringModel):
    
            def validation_step(self, batch, batch_idx):
                output = super().validation_step(batch, batch_idx)
                if self.trainer.sanity_checking:
                    self.log("val_loss", output["x"], prog_bar=True, logger=True)
                return output
    
            def training_step(self, batch, batch_idx):
                loss = super().training_step(batch, batch_idx)
                if batch_idx == 0:
                    assert self.trainer.logger_connector._progress_bar_metrics == {}
                    assert self.trainer.logger_connector._logged_metrics == {}
                    assert self.trainer.logger_connector._callback_metrics == {}
                self.log("train_loss", loss, prog_bar=True, logger=True)
                return loss
    
        trainer = Trainer(
            default_root_dir=tmpdir,
            max_epochs=1,
            limit_train_batches=1,
            limit_val_batches=2,
            num_sanity_val_steps=0,
        )
>       trainer.fit(TestModel())

ddsaf.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../pytorch_lightning/trainer/trainer.py:522: in fit
    self._run(model)
../pytorch_lightning/trainer/trainer.py:883: in _run
    self._dispatch()
../pytorch_lightning/trainer/trainer.py:925: in _dispatch
    self.accelerator.start_training(self)
../pytorch_lightning/accelerators/accelerator.py:96: in start_training
    self.training_type_plugin.start_training(trainer)
../pytorch_lightning/plugins/training_type/training_type_plugin.py:161: in start_training
    self._results = trainer.run_stage()
../pytorch_lightning/trainer/trainer.py:935: in run_stage
    return self._run_train()
../pytorch_lightning/trainer/trainer.py:983: in _run_train
    self.fit_loop.run()
../pytorch_lightning/loops/base.py:91: in run
    self.advance(*args, **kwargs)
../pytorch_lightning/loops/fit_loop.py:207: in advance
    epoch_output = self.epoch_loop.run(train_dataloader)
../pytorch_lightning/loops/base.py:92: in run
    self.on_advance_end()
../pytorch_lightning/loops/epoch/training_epoch_loop.py:156: in on_advance_end
    self._run_validation()
../pytorch_lightning/loops/epoch/training_epoch_loop.py:181: in _run_validation
    self.trainer.fit_loop.val_loop.run()
../pytorch_lightning/loops/base.py:97: in run
    output = self.on_run_end()
../pytorch_lightning/loops/dataloader/evaluation_loop.py:147: in on_run_end
    self.on_evaluation_end()
../pytorch_lightning/loops/dataloader/evaluation_loop.py:212: in on_evaluation_end
    self.trainer.call_hook("on_validation_end", *args, **kwargs)
../pytorch_lightning/trainer/trainer.py:1147: in call_hook
    trainer_hook(*args, **kwargs)
../pytorch_lightning/trainer/callback_hook.py:227: in on_validation_end
    callback.on_validation_end(self, self.lightning_module)
../pytorch_lightning/callbacks/model_checkpoint.py:279: in on_validation_end
    self.save_checkpoint(trainer)
../pytorch_lightning/callbacks/model_checkpoint.py:323: in save_checkpoint
    monitor_candidates = self._monitor_candidates(trainer, epoch=epoch, step=global_step)
../pytorch_lightning/callbacks/model_checkpoint.py:678: in _monitor_candidates
    monitor_candidates = deepcopy(trainer.logger_connector.callback_metrics)
../../../.conda/envs/pl/lib/python3.8/copy.py:146: in deepcopy
    y = copier(x, memo)
../../../.conda/envs/pl/lib/python3.8/copy.py:230: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
../../../.conda/envs/pl/lib/python3.8/copy.py:161: in deepcopy
    rv = reductor(4)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'loss': tensor(2.6577)}, drop_value = False

    def __getstate__(self, drop_value: bool = False) -> dict:
    
        def getstate(item: ResultMetric) -> dict:
            return item.__getstate__(drop_value=drop_value)
    
        items = apply_to_collection(dict(self), (ResultMetric, ResultMetricCollection), getstate)
>       return {"items": items, "meta": self.meta.__getstate__(), "_class": self.__class__.__name__}
E       AttributeError: 'NoneType' object has no attribute '__getstate__'

../pytorch_lightning/trainer/connectors/logger_connector/result.py:298: AttributeError
```

mind double check this?

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging_/test_train_loop_logging.py:717]: Hey @awaelchli,

Yes, got the same issue and resolved it with:

```
 if isinstance(value, ResultMetricCollection):
      value = dict(value.items())
```

We forgot to convert ResultMetricCollection to dictionaries.

Best,
T.C

**Review by SkafteNicki — APPROVED** [review]

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/logging_/test_train_loop_logging.py]: Done !

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -657,7 +657,10 @@ def _validate_monitor_key(self, trainer: 'pl.Trainer') -> None:
                 f" {list(metrics.keys())}. "
                 f"HINT: Did you call self.log('{self.monitor}', value) in the LightningModule?"
             )
-            raise MisconfigurationException(m)
+            if not trainer.fit_loop.val_loop._has_run:
+                warning_cache.warn(m)
+            else:
+                raise MisconfigurationException(m)
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:663]: @tchaton could you explain this check further? why does the val loop running need to be checked here for a warning vs an exception? relying on the val loop's property is confusing on first read since we support checkpointing during the training loop (every n batches)

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:663]: Hey @ananthsub, yes definitely !

ModelCheckpoint might rely on "val_loss" to checkpoint which is generated by validation_step.

We had flicky test relying on sanity_checking to assert the monitor value actually existed.

However, as the monitor metric doesn't exist until val_loop has been run, the ModelCheckpoint should just warn and skip checkpointing if the key is missing.

When val_loop has actually been run, then it should raise an exception if the metric key is missing (due to a typo or missing for step). 

If you rely on the metric generated by training_step, then the key will be present independently of  trainer.fit_loop.val_loop._has_run value and no warning or exception will be triggered.

Best,
T.C

Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:663]: > We had flicky test relying on sanity_checking to assert the monitor value actually existed.

But we don't save checkpoints during sanity checking at all. why'd it reach this point? is it possible for us to remove the `_has_run` property and its check here?
- having separate checks for train loop vs val loop is harder to maintain
- we can check if the trainer is sanity checking via the trainer property inside the callback
- `_has_run` is another form of progress tracking which sits outside the core tracking dataclasses, which we want to converge to

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:663]: I overlooked this.
I too think that the `_has_run` field here is a hack that we want to solve.
The `ModelCheckpoint` already tests against `trainer.sanity_checking`.  If there is still a reason for checking if the loop was run, we should rely on progress tracking.

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:663]: Yes, I agree. Once progress tracking has been added, this should be updated.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 4]
```diff
@@ -536,6 +536,10 @@ def metrics(self, on_step: bool) -> Dict[MetricSource, Dict[str, _METRIC]]:
             # extract forward_cache or computed from the ResultMetric. ignore when the output is None
             value = apply_to_collection(result_metric, ResultMetric, self._get_cache, on_step, include_none=False)
 
+            # convert metric collection to dict container.
+            if isinstance(value, ResultMetricCollection):
+                value = dict(value.items())
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:542]: Why is this necessary?

Also `dict(value)` woudl have been enough

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -269,6 +269,11 @@ def should_reset_tensors(self, fx: str) -> bool:
         return is_different_fx and is_first_batch
 
     def reset(self, metrics: Optional[bool] = None) -> None:
+        if self.trainer.sanity_checking:
+            # reset metrics
+            self._progress_bar_metrics = {}
+            self._logged_metrics = {}
+            self._callback_metrics = {}
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:276]: Two issues here:

1. We already call this function in `_run_sanity_check`. Instead of adding an `if` here, I woud've performed these calls there directly.

2. Why do we even need to recreate these? Wouldn't it be better to avoid logging anything in the first place during sanity check

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:276]: For 2) the sanity check is to make sure everything runs, so we should also internally record all the logs to make sure the user doesn't run into any logging errors. the only thing we would want to avoid is sending the logs to the logger object for saving.
